talktorials/11_online_apis/11c_quality_assessment.ipynb

```diff
@@ -0,0 +1,332 @@
+    "\n",
+    "1. Invoke the NGL viewer with the adequate representations\n",
+    "2. Build an interactive table of results (hint: use `ipywidgets.Select`)\n",
+    "3. Write an event handler that can communicate with the NGL Viewer: when the user clicks on a new entry, update the ligand in display, along with the surrounding residues. The ribbon should not need to be updated."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 35,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import time\n",
+    "import nglview as nv\n",
+    "from ipywidgets import AppLayout, Layout, Select"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The PDBQT file created by Vina contain several models, but `nglview` will only parse the first one. The workaround is simple: divide the file into individual models by splitting whenever an `ENDMDL` line is found."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def split_pdbqt(path):\n",
+    "    \"\"\"\n",
+    "    Split a multimodel PDBQT into separate files.\n",
+    "    \"\"\"\n",
+    "    files = []\n",
+    "    with open(path) as f:\n",
+    "        lines = []\n",
+    "        i = 0\n",
+    "        for line in f:\n",
+    "            lines.append(line)\n",
+    "            if line.strip() == 'ENDMDL':\n",
+    "                fn = f'data/results.{i}.pdbqt'\n",
+    "                with open(fn, 'w') as o:\n",
+    "                    o.write(''.join(lines))\n",
+    "                files.append(fn)\n",
+    "                i += 1\n",
+    "                lines = []\n",
+    "    return files"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The Vina output is a simple text file that contains the table of results. Parsing that table is relatively straightforward. We return a Pandas DataFrame for a simple visualization, if needed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def parse_output(out):\n",
+    "    \"\"\"\n",
+    "    Create a DataFrame out of the Vina output file\n",
+    "    \"\"\"\n",
+    "    with open(out) as f:\n",
+    "        data = []\n",
+    "        for line in f:\n",
+    "            if line.startswith('-----+'):\n",
+    "                line = next(f)\n",
+    "                while line.split()[0].isdigit():\n",
+    "                    index, *floats = line.split()\n",
+    "                    data.append([int(index)] + list(map(float, floats)))\n",
+    "                    line = next(f)\n",
+    "    return pd.DataFrame.from_records(data, \n",
+    "                                     columns=['Mode', 'Affinity (kcal/mol)', 'RMSD (l.b.)', 'RMSD (u.b.)'], \n",
+    "                                     exclude=['Mode'])"
+   ]
+  },
```

Comment:
I would add a cell that calls this function, in order to show the user how the vina output looks like and maybe discuss shortly the columns and values.

```
parse_output("data/vina.out")
```